File: src/lib/rule.ts
/**
 * Rule construction and the unit juggling the editor needs.
 *
 * The backend stores sizes in bytes and ages in seconds. People think in MB and
 * days, so conversions live here — pure and tested in rule.test.ts.
 */

import type { Action, ActionKind, Condition, ConditionKind, Rule } from './ipc';
import type { TranslationKey } from './i18n';

/** A blank rule, ready for the editor. */
export function newRule(order: number): Rule {
  return {
    id: crypto.randomUUID(),
    name: '',
    enabled: true,
    conditions: [],
    action: { type: 'Move', dest_template: '' },
    stop_on_match: true,
    order,
  };
}

/** A newly added condition of the given kind, with a sensible starting value. */
export function newCondition(kind: ConditionKind): Condition {
  switch (kind) {
    case 'Extension':
      return { type: 'Extension', value: '' };
    case 'Glob':
      return { type: 'Glob', value: '' };
    case 'Regex':
      return { type: 'Regex', value: '' };
    case 'MimeType':
      return { type: 'MimeType', value: '' };
    case 'SizeGreater':
      return { type: 'SizeGreater', value: 10 * 1024 * 1024 };
    case 'SizeSmaller':
      return { type: 'SizeSmaller', value: 1024 * 1024 };
    case 'OlderThan':
      return { type: 'OlderThan', value: 30 * 86400 };
    case 'NewerThan':
      return { type: 'NewerThan', value: 7 * 86400 };
    case 'MaxDepth':
      return { type: 'MaxDepth', value: 1 };
    case 'Duplicate':
      return { type: 'Duplicate' };
  }
}

/** Switches an action's kind, carrying the template across where it makes sense. */
export function changeActionKind(action: Action, kind: ActionKind): Action {
  const carried =
    action.type === 'Move' || action.type === 'Copy' || action.type === 'Unzip'
      ? action.dest_template
      : action.type === 'Rename'
        ? action.template
        : '';

  switch (kind) {
    case 'Move':
      return { type: 'Move', dest_template: carried };
    case 'Copy':
      return { type: 'Copy', dest_template: carried };
    case 'Unzip':
      return { type: 'Unzip', dest_template: carried };
    case 'Rename':
      return { type: 'Rename', template: carried };
    case 'RunCommand':
      return { type: 'RunCommand', command: '' };
    case 'Trash':
      return { type: 'Trash' };
  }
}

/** Replaces an action's template, whichever field the kind keeps it in. */
export function withTemplate(action: Action, template: string): Action {
  switch (action.type) {
    case 'Move':
    case 'Copy':
    case 'Unzip':
      return { ...action, dest_template: template };
    case 'Rename':
      return { ...action, template };
    default:
      return action;
  }
}

/** True when the kind carries a destination or name template. */
export function needsTemplate(kind: ActionKind): boolean {
  return kind === 'Move' || kind === 'Copy' || kind === 'Unzip' || kind === 'Rename';
}

// ---------------------------------------------------------------------------
// Sizes
// ---------------------------------------------------------------------------

export const SIZE_UNITS = [
  { label: 'B', bytes: 1 },
  { label: 'KB', bytes: 1024 },
  { label: 'MB', bytes: 1024 ** 2 },
  { label: 'GB', bytes: 1024 ** 3 },
] as const;

export type SizeUnit = (typeof SIZE_UNITS)[number]['label'];

/**
 * Splits a byte count into the largest unit that divides it cleanly.
 *
 * Bytes are in the list so there is always an exact answer: a size like 1536
 * that no larger unit divides evenly would otherwise have to be rounded, and
 * the editor would silently rewrite the rule on the next keystroke.
 */
export function splitSize(bytes: number): { amount: number; unit: SizeUnit } {
  for (const { label, bytes: size } of [...SIZE_UNITS].reverse()) {
    if (bytes >= size && bytes % size === 0) {
      return { amount: bytes / size, unit: label };
    }
  }
  return { amount: Math.max(0, Math.round(bytes)), unit: 'B' };
}

export function toBytes(amount: number, unit: SizeUnit): number {
  const found = SIZE_UNITS.find((u) => u.label === unit);
  return Math.max(0, Math.round(amount * (found?.bytes ?? 1)));
}

// ---------------------------------------------------------------------------
// Durations
// ---------------------------------------------------------------------------

export const AGE_UNITS = [
  { label: 'minutes', seconds: 60 },
  { label: 'hours', seconds: 3600 },
  { label: 'days', seconds: 86400 },
  { label: 'weeks', seconds: 604800 },
] as const;

export type AgeUnit = (typeof AGE_UNITS)[number]['label'];

export function splitAge(seconds: number): { amount: number; unit: AgeUnit } {
  for (const { label, seconds: size } of [...AGE_UNITS].reverse()) {
    if (seconds >= size && seconds % size === 0) {
      return { amount: seconds / size, unit: label };
    }
  }
  return { amount: Math.max(1, Math.round(seconds / 60)), unit: 'minutes' };
}

export function toSeconds(amount: number, unit: AgeUnit): number {
  const found = AGE_UNITS.find((u) => u.label === unit);
  return Math.max(0, Math.round(amount * (found?.seconds ?? 60)));
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

/**
 * Why a rule can't be saved yet, as a translation key, or null when it's fine.
 * Deliberately permissive: a rule with no conditions is legal and matches
 * everything, which is a real (if bold) thing to want.
 */
export function ruleProblem(rule: Rule): TranslationKey | null {
  if (!rule.name.trim()) return 'editor.needName';

  for (const condition of rule.conditions) {
    if ('value' in condition && typeof condition.value === 'string' && !condition.value.trim()) {
      return 'editor.emptyCondition';
    }
    if (condition.type === 'Regex') {
      try {
        new RegExp(condition.value);
      } catch {
        return 'editor.badRegex';
      }
    }
  }

  switch (rule.action.type) {
    case 'Move':
    case 'Copy':
    case 'Unzip':
      if (!rule.action.dest_template.trim()) return 'editor.needDestination';
      break;
    case 'Rename':
      if (!rule.action.template.trim()) return 'editor.needTemplate';
      break;
    case 'RunCommand':
      if (!rule.action.command.trim()) return 'editor.needCommand';
      break;
    case 'Trash':
      break;
  }

  return null;
}

File: src-tauri/src/core/rule.rs
// Rule model and serialization.

use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use uuid::Uuid;

/// A file organization rule.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Rule {
    pub id: Uuid,
    pub name: String,
    pub enabled: bool,
    pub conditions: Vec<Condition>,
    pub action: Action,
    pub stop_on_match: bool,
    pub order: i32,
}

/// Conditions that a file must satisfy.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "value")]
pub enum Condition {
    Extension(String),
    Glob(String),
    Regex(String),
    SizeGreater(u64),
    SizeSmaller(u64),
    OlderThan(i64),
    NewerThan(i64),
    MimeType(String),
    Duplicate,
    MaxDepth(u32),
}

/// Action to perform when a rule matches.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum Action {
    Move { dest_template: String },
    Copy { dest_template: String },
    Rename { template: String },
    Trash,
    Unzip { dest_template: String },
    RunCommand { command: String },
}

impl Rule {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            id: Uuid::new_v4(),
            name: name.into(),
            enabled: true,
            conditions: Vec::new(),
            action: Action::Move {
                dest_template: String::new(),
            },
            stop_on_match: true,
            order: 0,
        }
    }
}

/// A planned file operation (dry-run result).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlannedOperation {
    pub id: Uuid,
    pub source: PathBuf,
    pub destination: Option<PathBuf>,
    pub action_type: ActionType,
    pub rule_id: Uuid,
    pub rule_name: String,
    pub selected: bool,
}

/// Simplified action type for UI display.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ActionType {
    Move,
    Copy,
    Rename,
    Trash,
    Unzip,
    RunCommand,
}

impl From<&Action> for ActionType {
    fn from(action: &Action) -> Self {
        match action {
            Action::Move { .. } => ActionType::Move,
            Action::Copy { .. } => ActionType::Copy,
            Action::Rename { .. } => ActionType::Rename,
            Action::Trash => ActionType::Trash,
            Action::Unzip { .. } => ActionType::Unzip,
            Action::RunCommand { .. } => ActionType::RunCommand,
        }
    }
}
